funcs/azure.go

```diff
@@ -0,0 +1,79 @@
+package funcs
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"sync"
+
+	"github.com/hairyhenderson/gomplate/v4/azure"
+)
+
+// AzureNS - the azure namespace
+//
+// Deprecated: don't use
+func AzureNS() *AzureFuncs {
+	return &AzureFuncs{azureOpts: azure.GetClientOptions()}
+}
+
+// AddAzureFuncs -
+//
+// Deprecated: use [CreateAzureFuncs] instead
+func AddAzureFuncs(f map[string]interface{}) {
+	for k, v := range CreateAzureFuncs(context.Background()) {
+		f[k] = v
+	}
+}
+
+// CreateAzureFuncs -
+func CreateAzureFuncs(ctx context.Context) map[string]interface{} {
+	ns := &AzureFuncs{
+		ctx:       ctx,
+		azureOpts: azure.GetClientOptions(),
+	}
+	return map[string]interface{}{
+		"azure": func() interface{} { return ns },
+	}
+}
+
+// AzureFuncs -
+type AzureFuncs struct {
+	ctx context.Context
+
+	meta      *azure.MetaClient
+	metaInit  sync.Once
+	azureOpts azure.ClientOptions
+}
+
+// Meta -
+func (a *AzureFuncs) Meta(args ...string) (string, error) {
+	if len(args) == 0 || len(args) > 3 {
+		return "", fmt.Errorf("wrong number of args: wanted 1, 2 or 3, got %d", len(args))
+	}
+	key := args[0]
+	format := "text"
+	def := args[1:]
+	apiVersion := "2021-12-13"
+
+	if len(args) >= 2 {
+		if args[1] == "json" || args[1] == "text" {
+			format = args[1]
+			def = args[2:]
+		}
+	}
+	if len(args) >= 3 {
+		if found, _ := regexp.MatchString(`^\d{4}-\d{2}-\d{2}$`, args[2]); found {
+			apiVersion = args[2]
```

Comment:
I'm not sure providing the API version as an argument makes a lot of sense - this seems like something that users will very rarely need to set to anything other than the latest.

And if they do need to change it I'd expect they'd need to change it to the same version for all calls in the template.

How about an environment variable for this instead, like `AZURE_API_VERSION`?

Reply:
I added the API version here to make this function more future proof. The current default is the lastest, but it could be changed in future. 

> AZURE_API_VERSION

There is no global API Version at Azure. Each service can have a own API Version while the extact API Version may not exists on a different API.

It would work, since there is on IMDS for now. But if someone add Azure Key Vault, the AZURE_API_VERSION cant be re-used there.

Reply:
fair - then `AZURE_IMDS_API_VERSION` maybe...

> But if someone add Azure Key Vault [...]

yup, but I'd think it would make way more sense to add that as a datasource.

Which brings up [this conversation](https://github.com/hairyhenderson/gomplate/pull/891#issuecomment-656899540) again... Does `azure.Meta` _need_ to be a function? could there be an `azure+meta://` datasource instead?

Reply:
From my point of view, both is fine. At datasource could reduce the calls to the IMDS service too.

But from user perspective, there is no difference between

```
{{- defineDatasource "imds" "azure+meta://instance" }}
Hello {{ (ds "imds").resourceId }}
```

or

```
Hello {{ azure.meta "instance/resourceId" "text" }}
```